From 03750ed13a169bb812ddfbb83906d6717d802d23 Mon Sep 17 00:00:00 2001
From: Guilhem Bichot <guilhem.bichot@contributor>

Date: Wed, 24 Sep 2025 11:15:42 +0200
Subject: [PATCH] decrease the memory footprint of session
 plan cache.

tests session_plan_cache_debug in main and dstore_main suite are 90%
identical (the only difference is that the dstore_main one has an
additional test at its very end), so let's make an include file to
reduce maintenance effort.
---
 .../include/session_plan_cache_debug.inc      | 131 ++++++++++++++++++
 .../t/session_plan_cache_debug.test           | 128 +----------------
 mysql-test/t/session_plan_cache_debug.test    | 112 +--------------
 3 files changed, 135 insertions(+), 236 deletions(-)
 create mode 100644 mysql-test/include/session_plan_cache_debug.inc

diff --git a/mysql-test/include/session_plan_cache_debug.inc b/mysql-test/include/session_plan_cache_debug.inc
new file mode 100644
index 00000000000..488b4d28060
--- /dev/null
+++ b/mysql-test/include/session_plan_cache_debug.inc
@@ -0,0 +1,131 @@
+--source include/have_debug.inc
+--source include/have_dstore.inc
+
+# Tests of session-level plan cache which require a debug binary
+
+# We will be displaying global status variables which FLUSH STATUS does not reset (like Cached_plan_invalidations); if mysqld has already executed some other mtr tests and they used plan cache, this status will contain effects from those other tests, so restart:
+--source include/restart_mysqld.inc
+
+# PLAN_CACHE_PORT: here some sections moved from session_plan_cache.test
+
+set rds_plan_cache=ON;
+
+--echo "test optimizer trace for plan cache"
+flush status;
+show session status like "cached_plan%";
+create table t1(a int, b int);
+insert into t1 values(1,2),(3,4);
+prepare stmt from 'select * from t1 where a > ?';
+set @a = 1;
+set optimizer_trace='enabled=on';
+execute stmt using @a;
+show session status like "cached_plan%";
+execute stmt using @a;
+show session status like "cached_plan%";
+
+set debug='+d,apply_cached_plan_fail';
+execute stmt using @a;
+--replace_regex /  \"cost\".*// /  \"cost_for_plan\".*//
+select * from information_schema.optimizer_trace;
+show session status like "cached_plan%";
+set optimizer_trace='enabled=off';
+drop table t1;
+set debug='-d,apply_cached_plan_fail';
+
+--echo # BUGFIX : mysqld crash, keyword:plan_cache::exec_cached_plan.
+
+CREATE TABLE `sbtest` (
+	`id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+	`k` int(10) UNSIGNED NOT NULL DEFAULT '0',
+	`c` char(120) NOT NULL DEFAULT '',
+	`pad` char(60) NOT NULL DEFAULT '',
+	PRIMARY KEY (`id`),
+	KEY `k` (`k`)
+) engine=innodb;
+
+# PLAN_CACHE_PORT we use innodb above, not dstore; because we want to
+# inject an error at the point ("pq_dup_tabs_error0") where the
+# condition attached to the table is cloned; in dstore there is no
+# such condition (it was moved to ICP).
+
+DELIMITER |;
+CREATE PROCEDURE InsertRandomData()
+BEGIN
+    DECLARE i INT DEFAULT 1;
+    DECLARE random_value INT;
+    DECLARE repeated_values_count INT DEFAULT 200;
+
+    SET @@session.rand_seed1 = 12345;
+    SET @@session.rand_seed2 = 67890;
+
+    WHILE i <= 1000 DO
+        IF i <= repeated_values_count THEN
+            SET random_value = FLOOR(1 + (RAND() * 100));
+        ELSE
+            SET random_value = FLOOR(101 + (RAND() * 900));
+        END IF;
+
+        INSERT INTO sbtest (`k`, `c`, `pad`)
+        VALUES (i, CONCAT('value_', random_value), CONCAT('pad_', i));
+
+        SET i = i + 1;
+    END WHILE;
+END|
+DELIMITER ;|
+
+CALL InsertRandomData();
+
+PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
+
+SET @n = 10;
+SET @m = 15;
+EXECUTE stmt USING @n, @m;
+EXECUTE stmt USING @n, @m;
+
+SET debug= "+d,plan_cache_debug_test_quick_select_fail";
+EXECUTE stmt USING @n, @m;
+SET debug="-d,plan_cache_debug_test_quick_select_fail";
+
+DROP PROCEDURE InsertRandomData;
+DEALLOCATE PREPARE stmt;
+
+# PLAN_CACHE_PORT New sections here.
+
+PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
+# Simulate OOM failure which happens when caching the plan
+SET session debug="+d,pq_dup_tabs_error0";
+flush status;
+EXECUTE stmt USING @n, @m;
+EXECUTE stmt USING @n, @m;
+EXECUTE stmt USING @n, @m;
+# Nothing should be cached
+show session status like "cached_plan%";
+SET session debug="-d,pq_dup_tabs_error0";
+PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
+EXECUTE stmt USING @n, @m;
+EXECUTE stmt USING @n, @m;
+EXECUTE stmt USING @n, @m;
+# Now caching should work
+show session status like "cached_plan%";
+DEALLOCATE PREPARE stmt;
+
+if ($engine_is_dstore)
+{
+--echo #
+--echo # BUGFIX after enable rds_plan_cache, using the sysbench range query model(select_random_ranges.lua), the performance fluctuates sharply and the performance decreases by 5% to 25%
+--echo #
+
+# Even though EXPLAIN reports 'Using index', the actual reading of the constant table read the full row. Make sure this does not happen anymore:
+alter table sbtest engine=dstore;
+set debug='+d,dstore_must_read_only_index';
+explain SELECT id FROM sbtest WHERE id=100;
+prepare s from 'SELECT id FROM sbtest WHERE id=?';
+set @a=100;
+execute s using @a;
+execute s using @a;
+execute s using @a;
+show session status like "cached_plan%";
+set debug='-d,dstore_must_read_only_index';
+}
+
+DROP TABLE sbtest;
diff --git a/mysql-test/suite/dstore_main/t/session_plan_cache_debug.test b/mysql-test/suite/dstore_main/t/session_plan_cache_debug.test
index fd0062ca268..542c2d2cc9c 100644
--- a/mysql-test/suite/dstore_main/t/session_plan_cache_debug.test
+++ b/mysql-test/suite/dstore_main/t/session_plan_cache_debug.test
@@ -1,128 +1,4 @@
 --source include/have_debug.inc
 --source include/have_dstore.inc
-
-# Tests of session-level plan cache which require a debug binary
-
-# We will be displaying global status variables which FLUSH STATUS does not reset (like Cached_plan_invalidations); if mysqld has already executed some other mtr tests and they used plan cache, this status will contain effects from those other tests, so restart:
---source include/restart_mysqld.inc
-
-# PLAN_CACHE_PORT: here some sections moved from session_plan_cache.test
-
-set rds_plan_cache=ON;
-
---echo "test optimizer trace for plan cache"
-flush status;
-show session status like "cached_plan%";
-create table t1(a int, b int);
-insert into t1 values(1,2),(3,4);
-prepare stmt from 'select * from t1 where a > ?';
-set @a = 1;
-set optimizer_trace='enabled=on';
-execute stmt using @a;
-show session status like "cached_plan%";
-execute stmt using @a;
-show session status like "cached_plan%";
-
-set debug='+d,apply_cached_plan_fail';
-execute stmt using @a;
---replace_regex /  \"cost\".*// /  \"cost_for_plan\".*//
-select * from information_schema.optimizer_trace;
-show session status like "cached_plan%";
-set optimizer_trace='enabled=off';
-drop table t1;
-set debug='-d,apply_cached_plan_fail';
-
---echo # BUGFIX : mysqld crash, keyword:plan_cache::exec_cached_plan.
-
-CREATE TABLE `sbtest` (
-	`id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
-	`k` int(10) UNSIGNED NOT NULL DEFAULT '0',
-	`c` char(120) NOT NULL DEFAULT '',
-	`pad` char(60) NOT NULL DEFAULT '',
-	PRIMARY KEY (`id`),
-	KEY `k` (`k`)
-) engine=innodb;
-
-# PLAN_CACHE_PORT we use innodb above, not dstore; because we want to
-# inject an error at the point ("pq_dup_tabs_error0") where the
-# condition attached to the table is cloned; in dstore there is no
-# such condition (it was moved to ICP).
-
-DELIMITER |;
-CREATE PROCEDURE InsertRandomData()
-BEGIN
-    DECLARE i INT DEFAULT 1;
-    DECLARE random_value INT;
-    DECLARE repeated_values_count INT DEFAULT 200;
-
-    SET @@session.rand_seed1 = 12345;
-    SET @@session.rand_seed2 = 67890;
-
-    WHILE i <= 1000 DO
-        IF i <= repeated_values_count THEN
-            SET random_value = FLOOR(1 + (RAND() * 100));
-        ELSE
-            SET random_value = FLOOR(101 + (RAND() * 900));
-        END IF;
-
-        INSERT INTO sbtest (`k`, `c`, `pad`)
-        VALUES (i, CONCAT('value_', random_value), CONCAT('pad_', i));
-
-        SET i = i + 1;
-    END WHILE;
-END|
-DELIMITER ;|
-
-CALL InsertRandomData();
-
-PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
-
-SET @n = 10;
-SET @m = 15;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-
-SET debug= "+d,plan_cache_debug_test_quick_select_fail";
-EXECUTE stmt USING @n, @m;
-SET debug="-d,plan_cache_debug_test_quick_select_fail";
-
-DROP PROCEDURE InsertRandomData;
-DEALLOCATE PREPARE stmt;
-
-# PLAN_CACHE_PORT New sections here.
-
-PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
-# Simulate OOM failure which happens when caching the plan
-SET session debug="+d,pq_dup_tabs_error0";
-flush status;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-# Nothing should be cached
-show session status like "cached_plan%";
-SET session debug="-d,pq_dup_tabs_error0";
-PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-# Now caching should work
-show session status like "cached_plan%";
-DEALLOCATE PREPARE stmt;
-
---echo #
---echo # BUGFIX after enable rds_plan_cache, using the sysbench range query model(select_random_ranges.lua), the performance fluctuates sharply and the performance decreases by 5% to 25%
---echo #
-
-# Even though EXPLAIN reports 'Using index', the actual reading of the constant table read the full row. Make sure this does not happen anymore:
-alter table sbtest engine=dstore;
-set debug='+d,dstore_must_read_only_index';
-explain SELECT id FROM sbtest WHERE id=100;
-prepare s from 'SELECT id FROM sbtest WHERE id=?';
-set @a=100;
-execute s using @a;
-execute s using @a;
-execute s using @a;
-show session status like "cached_plan%";
-set debug='-d,dstore_must_read_only_index';
-
-DROP TABLE sbtest;
+let $engine_is_dstore=1;
+--source include/session_plan_cache_debug.inc
diff --git a/mysql-test/t/session_plan_cache_debug.test b/mysql-test/t/session_plan_cache_debug.test
index dbd9679aadb..0c43c6d5166 100644
--- a/mysql-test/t/session_plan_cache_debug.test
+++ b/mysql-test/t/session_plan_cache_debug.test
@@ -1,112 +1,4 @@
 --source include/have_debug.inc
 --source include/not_have_only_dstore.inc
-
-# Tests of session-level plan cache which require a debug binary
-
-# We will be displaying global status variables which FLUSH STATUS does not reset (like Cached_plan_invalidations); if mysqld has already executed some other mtr tests and they used plan cache, this status will contain effects from those other tests, so restart:
---source include/restart_mysqld.inc
-
-# PLAN_CACHE_PORT: here some sections moved from session_plan_cache.test
-
-set rds_plan_cache=ON;
-
---echo "test optimizer trace for plan cache"
-flush status;
-show session status like "cached_plan%";
-create table t1(a int, b int);
-insert into t1 values(1,2),(3,4);
-prepare stmt from 'select * from t1 where a > ?';
-set @a = 1;
-set optimizer_trace='enabled=on';
-execute stmt using @a;
-show session status like "cached_plan%";
-execute stmt using @a;
-show session status like "cached_plan%";
-
-set debug='+d,apply_cached_plan_fail';
-execute stmt using @a;
---replace_regex /  \"cost\".*// /  \"cost_for_plan\".*//
-select * from information_schema.optimizer_trace;
-show session status like "cached_plan%";
-set optimizer_trace='enabled=off';
-drop table t1;
-set debug='-d,apply_cached_plan_fail';
-
---echo # BUGFIX : mysqld crash, keyword:plan_cache::exec_cached_plan.
-
-CREATE TABLE `sbtest` (
-	`id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
-	`k` int(10) UNSIGNED NOT NULL DEFAULT '0',
-	`c` char(120) NOT NULL DEFAULT '',
-	`pad` char(60) NOT NULL DEFAULT '',
-	PRIMARY KEY (`id`),
-	KEY `k` (`k`)
-) engine=innodb;
-
-# PLAN_CACHE_PORT we use innodb above, not dstore; because we want to
-# inject an error at the point ("pq_dup_tabs_error0") where the
-# condition attached to the table is cloned; in dstore there is no
-# such condition (it was moved to ICP).
-
-DELIMITER |;
-CREATE PROCEDURE InsertRandomData()
-BEGIN
-    DECLARE i INT DEFAULT 1;
-    DECLARE random_value INT;
-    DECLARE repeated_values_count INT DEFAULT 200;
-
-    SET @@session.rand_seed1 = 12345;
-    SET @@session.rand_seed2 = 67890;
-
-    WHILE i <= 1000 DO
-        IF i <= repeated_values_count THEN
-            SET random_value = FLOOR(1 + (RAND() * 100));
-        ELSE
-            SET random_value = FLOOR(101 + (RAND() * 900));
-        END IF;
-
-        INSERT INTO sbtest (`k`, `c`, `pad`)
-        VALUES (i, CONCAT('value_', random_value), CONCAT('pad_', i));
-
-        SET i = i + 1;
-    END WHILE;
-END|
-DELIMITER ;|
-
-CALL InsertRandomData();
-
-PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
-
-SET @n = 10;
-SET @m = 15;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-
-SET debug= "+d,plan_cache_debug_test_quick_select_fail";
-EXECUTE stmt USING @n, @m;
-SET debug="-d,plan_cache_debug_test_quick_select_fail";
-
-DROP PROCEDURE InsertRandomData;
-DEALLOCATE PREPARE stmt;
-
-# PLAN_CACHE_PORT New sections here.
-
-PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
-# Simulate OOM failure which happens when caching the plan
-SET session debug="+d,pq_dup_tabs_error0";
-flush status;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-# Nothing should be cached
-show session status like "cached_plan%";
-SET session debug="-d,pq_dup_tabs_error0";
-PREPARE stmt FROM 'SELECT DISTINCT c FROM sbtest WHERE id BETWEEN ? AND ? ORDER BY c';
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-EXECUTE stmt USING @n, @m;
-# Now caching should work
-show session status like "cached_plan%";
-DEALLOCATE PREPARE stmt;
-
-DROP TABLE sbtest;
+let $engine_is_dstore=0;
+--source include/session_plan_cache_debug.inc
-- 
2.23.0

